js/components/pages/PassagePage.react.js

```diff
@@ -5,60 +5,179 @@ import React, { Component } from 'react';
 import { connect } from 'react-redux';
 import { Link } from 'react-router';
 
-import { asyncNextVerse, asyncPreviousVerse, asyncChangeMode } from '../../actions/AppActions';
-import { MODES, SINGLE_MODE, MULTI_MODE } from '../../constants/AppConstants';
+import { asyncNavigateNext, asyncNavigatePrevious, asyncChangeMode, asyncChangeRecall } from '../../actions/AppActions';
+import { VERSE_MODE, SEGMENT_MODE, CHAPTER_MODE, RECALL_STAGES } from '../../constants/AppConstants';
 
 import Verse from '../Verse.react';
+import AudioPlayer from '../AudioPlayer.react';
+import Swipeable from 'react-swipeable';
 
 // use named export for unconnected component (unit tests)
 export class PassagePage extends Component {
   render() {
     const dispatch = this.props.dispatch;
-    const { activeVerse, totalVerses, verses, mode } = this.props.data;
-
-    let nextButton = ''
-    let previousButton = ''
-    if (mode === SINGLE_MODE) {
-      previousButton = <button
-        title="Previous"
-        disabled={activeVerse == 0}
-        className="previous"
-        onClick={() => { dispatch(asyncPreviousVerse()) }}
-      >Previous</button>
-      nextButton = <button
-        title="Next"
-        disabled={activeVerse >= totalVerses - 1}
-        className="next"
-        onClick={() => { dispatch(asyncNextVerse()) }}
-      >Next</button>
+    const { active, lowerBound, upperBound, verses, mode, recallStage, isAudioPlaying } = this.props.data;
+
+    /**
+     * Take into account the current mode, and determine if we can navigation backward.
+     *
+     * @TODO make this better than a simple index check (which is only good for verse mode)
+     *
+     * @return boolean Whether or not we can navigate backward
+     */
+    function canNavigatePrevious() {
```

Comment:
Ideally, I'd like to keep these types of calculations outside of the components as it would make it easier to test.

One thought I had it make the reducer able to handle the action all the time.  That way if you were already in the first verse it would do nothing.  It would at least simplify our initial version.